### PR TITLE
Added support for customizable termination timeout

### PR DESCRIPTION
**Pull request by chtyim**
- Added a new terminate(long, TimeUnit) method in ServiceController

### CODE REVIEW

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java [hunk 1]
```diff
@@ -210,21 +210,22 @@ protected void doStartUp() {
...
     @Override
     protected void doShutDown() {
-      // Wait for sometime for the container to stop
-      // TODO: Use configurable value for stop time (TWILL-192)
-      int maxWaitSecs = Constants.APPLICATION_MAX_STOP_SECONDS - 30;
-      maxWaitSecs = maxWaitSecs < 15 ? 15 : maxWaitSecs;
+      // Wait for sometime for the container to stop. Subtracts 15 seconds from the timeout to accommodate for
+      // the message propagation delay.
+      long timeoutMillis = getTerminationTimeoutMillis(Constants.APPLICATION_MAX_STOP_SECONDS, TimeUnit.SECONDS);
+      timeoutMillis = timeoutMillis < 30L ? timeoutMillis : timeoutMillis - 15L;
```
Comment (yaojiefeng) [inline twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: The unit of the timeout here changes from secs -> millis, should we minus 15000L here?

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java [hunk 2]
```diff
@@ -210,21 +210,22 @@ protected void doStartUp() {
...
           return;
         }
       } catch (Exception e) {
         LOG.error("Got exception while trying to stop runnable {}, instance {}", runnable, instanceId, e);
       }
       // Container has not shutdown even after maxWaitSecs after sending stop message,
       // we'll need to kill the container
-      LOG.warn("Killing runnable {}, instance {} after waiting {} secs", runnable, instanceId, maxWaitSecs);
-      killAndWait(maxWaitSecs);
+      LOG.warn("Killing runnable {}, instance {} after waiting {} secs", runnable, instanceId, timeoutMillis);
```
Comment (yaojiefeng) [inline twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: same here, secs -> millis

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java [hunk 3]
```diff
@@ -210,21 +210,22 @@ protected void doStartUp() {
...
-      int maxWaitSecs = Constants.APPLICATION_MAX_STOP_SECONDS - 30;
-      maxWaitSecs = maxWaitSecs < 15 ? 15 : maxWaitSecs;
+      // Wait for sometime for the container to stop. Subtracts 15 seconds from the timeout to accommodate for
+      // the message propagation delay.
+      long timeoutMillis = getTerminationTimeoutMillis(Constants.APPLICATION_MAX_STOP_SECONDS, TimeUnit.SECONDS);
+      timeoutMillis = timeoutMillis < 30L ? timeoutMillis : timeoutMillis - 15L;
+
       try {
-        if (Uninterruptibles.awaitUninterruptibly(shutdownLatch, maxWaitSecs, TimeUnit.SECONDS)) {
+        if (Uninterruptibles.awaitUninterruptibly(shutdownLatch, timeoutMillis, TimeUnit.SECONDS)) {
```
Comment (yaojiefeng) [inline twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: should it be `TimeUnit.MILLISECONDS`?

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java [hunk 4]
```diff
@@ -210,21 +210,22 @@ protected void doStartUp() {
...
         }
       } catch (Exception e) {
         LOG.error("Got exception while trying to stop runnable {}, instance {}", runnable, instanceId, e);
       }
       // Container has not shutdown even after maxWaitSecs after sending stop message,
       // we'll need to kill the container
-      LOG.warn("Killing runnable {}, instance {} after waiting {} secs", runnable, instanceId, maxWaitSecs);
-      killAndWait(maxWaitSecs);
+      LOG.warn("Killing runnable {}, instance {} after waiting {} secs", runnable, instanceId, timeoutMillis);
+      killAndWait(timeoutMillis);
```
Comment (albertshau) [inline twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: shouldn't this be translated to seconds first?

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java (hunk 1 above)
Comment (chtyim) [reply twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: Actually the timeout should just be in seconds as the rest of the logic are all in seconds.

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java (hunk 2 above)
Comment (chtyim) [reply twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: Changed to use seconds.

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java (hunk 3 above)
Comment (chtyim) [reply twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: Changed to use seconds.

File: twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java (hunk 4 above)
Comment (chtyim) [reply twill-core/src/main/java/org/apache/twill/internal/TwillContainerLauncher.java]: Changed to use seconds.

**Review by albertshau — APPROVED** [review]

**Review by yaojiefeng — APPROVED** [review]

**Comment by snehalakshmisha** [comment]
Thanks for working on this PR.

**Pull request merged**